File: examples/tree-simple.c
#include <stdio.h>
#include "Fl_C.h"
#include "Fl_EnumerationsC.h"
#include "Fl_Types.h"

void TreeCallback(fl_Widget w, void* data){
  fl_Tree tree = (fl_Tree)w;
  fl_Tree_Item item = (fl_Tree_Item)Fl_Tree_callback_item(tree);
  if (! item ) return;
  switch( Fl_Tree_callback_reason(tree)) {
    case FL_TREE_REASON_SELECTED: {
      char pathname[256];
      Fl_Tree_item_pathname(tree,pathname, sizeof(pathname), item);
      printf("TreeCallback: Item selected='%s', Full pathname='%s'\n", (const char*)(Fl_Tree_Item_label(item)), pathname);
      break;
    }
    case FL_TREE_REASON_DESELECTED:
      printf("TreeCallback: Item '%s' deselected\n", (const char*)(Fl_Tree_Item_label(item)));
      break;
    case FL_TREE_REASON_OPENED:
      printf("TreeCallback: Item '%s' opened\n", (const char*)(Fl_Tree_Item_label(item)));
      break;
    case FL_TREE_REASON_CLOSED:
      printf("TreeCallback: Item '%s' closed\n", (const char*)(Fl_Tree_Item_label(item)));
    default:
      break;
  }
}
int main(int argc, char** argv) {
  Fl_set_scheme("gtk+");
  fl_Window window = (fl_Window)Fl_Window_New_WithTitle(250,400,"Simple Tree");
  Fl_Window_begin(window);
  fl_Tree tree = (fl_Tree)Fl_Tree_New(10,10,Fl_Window_w(window)-20,Fl_Window_h(window)-20);
  Fl_Tree_showroot(tree,0);

  Fl_Tree_add_with_path(tree,"Flintstones/Fred");
  Fl_Tree_add_with_path(tree,"Flintstones/Wilma");
  Fl_Tree_add_with_path(tree,"Flintstones/Pebbles");
  Fl_Tree_add_with_path(tree,"Simpsons/Homer");
  Fl_Tree_add_with_path(tree,"Simpsons/Marge");
  Fl_Tree_add_with_path(tree,"Simpsons/Bart");
  Fl_Tree_add_with_path(tree,"Simpsons/Lisa");
  Fl_Tree_add_with_path(tree,"Pathnames/\\/bin");		// front slashes
  Fl_Tree_add_with_path(tree,"Pathnames/\\/usr\\/sbin");
  Fl_Tree_add_with_path(tree,"Pathnames/C:\\\\Program Files");	// backslashes
  Fl_Tree_add_with_path(tree,"Pathnames/C:\\\\Documents and Settings");

  Fl_Tree_close_with_path(tree,"Simpsons");
  Fl_Tree_close_with_path(tree,"Pathnames");
  Fl_Tree_set_callback(tree,TreeCallback);
  Fl_Window_end(window);
  Fl_Window_show(window);
  return Fl_run();
}

File: fltkhs-demos/src/Examples/tree-simple.hs
{-# LANGUAGE OverloadedStrings #-}
module Main where
import qualified Graphics.UI.FLTK.LowLevel.FL as FL
import Graphics.UI.FLTK.LowLevel.Fl_Types
import Graphics.UI.FLTK.LowLevel.FLTKHS
import qualified Data.Text as T
treeCallback :: Ref Tree -> IO ()
treeCallback tree' = do
  (Just item') <- getCallbackItem tree'
  reason' <- getCallbackReason tree'
  label' <- getLabel item'
  case reason' of
    TreeReasonSelected -> do
      (Just path') <- itemPathname tree' item'
      print $ "TreeCallback: Item selected =" ++ (T.unpack label') ++ "Full pathname=" ++ (T.unpack path') ++ "\n"
    TreeReasonDeselected ->
      print $ "TreeCallback: Item deselected =" ++ (T.unpack label') ++ "\n"
    TreeReasonOpened ->
      print $ "TreeCallback: Item opened =" ++ (T.unpack label') ++ "\n"
    TreeReasonClosed ->
      print $ "TreeCallback: Item closed =" ++ (T.unpack label') ++ "\n"
    _ -> print ("" :: String)
main :: IO ()
main = do
  _ <- FL.setScheme "gtk+"
  window <- windowNew (toSize (250,400)) Nothing (Just "Simple Tree")
  begin window
  (Width windowWidth') <- getW window
  (Height windowHeight') <- getH window
  tree <- treeNew (toRectangle (10,10,windowWidth' - 20,windowHeight' - 20)) Nothing
  setShowroot tree False
  _ <- add tree "Flintstones/Fred"
  _ <- add tree "Flintstones/Wilma"
  _ <- add tree "Flintstones/Pebbles"
  _ <- add tree "Simpsons/Homer"
  _ <- add tree "Simpsons/Marge"
  _ <- add tree "Simpsons/Bart"
  _ <- add tree "Simpsons/Lisa"
  _ <- add tree "Pathnames/\\/bin"
  _ <- add tree "Pathnames/\\/usr\\/sbin"
  _ <- add tree "Pathnames/C:\\\\Program Files"
  _ <- add tree "Pathnames/C:\\\\Documents and Settings"
  close tree (TreeItemNameLocator (TreeItemName "Simpsons"))
  close tree (TreeItemNameLocator (TreeItemName "Pathnames"))
  setCallback tree treeCallback
  end window
  showWidget window
  _ <- FL.run
  return ()
